### ISSUE TITLE
Question regarding verify_cert_identity when using the GitHub Actions credential

### ISSUE THREAD

**Issue by Lucas-C** [sigstore/gh-action-sigstore-python#158]
Hi!

I'm trying to setup `sigstore` on https://github.com/py-pdf/fpdf2

I just switched to `v3.0.0` and discovered the new `upload-signing-artifacts` setting, which I'm going to enable today, so that for the next release sigstore bundles can be downloaded as GitHub job artifacts.

PR: https://github.com/py-pdf/fpdf2/pull/1331

I also wanted to enable [verify](https://github.com/sigstore/gh-action-sigstore-python?tab=readme-ov-file#verify), but I'm a bit hesitant regarding the value to provide for `verify-cert-identity` when using this GitHub Action and the OIDC token used for signing comes from GitHub?

If I'm not mistaken, a similar `--cert-identity` value will have to be provided to `sigstore verify identity` when end users will want to [verify a `fpdf2` package version against a sigstore bundle](https://github.com/sigstore/sigstore-python#verifying-against-a-bundle).

And again, I'm not sure which value to provide, and it does not seem displayed in the GitHub Action log.

Could you please provide some guidance regarding this? 🙂

**Comment by Lucas-C** [comment]
Thank you very much.

I'm going to experiment with `sigstore verify github` command to figure how to perform a verification based on the data publish on Pypi.

On thing I did not figure out yet is: where to retrieve a `.sigstore.json` or `.bundle` file from Pypi?
One such file seems to be required by `sigstore verify`

**Comment by Lucas-C** [comment]
I tried to dig a little further regarding this...

I found a comment from 18 months ago on https://github.com/sigstore/sigstore-python that implies that `.crt` / `.sig` files are (were?) uploaded onto Pypi: https://github.com/sigstore/sigstore-python/issues/126#issuecomment-1151030983

However I was not able to download them for the latest `fpdf2` release:
```
$ wget https://github.com/actions/python-versions/releases/download/3.13.1-12154081405/python-3.13.1-linux-22.04-x64.tar.gz.crt
2024-12-27 08:58:47 ERROR 404: Not Found.

$ $ wget https://github.com/actions/python-versions/releases/download/3.13.1-12154081405/python-3.13.1-linux-22.04-x64.tar.gz.sig
2024-12-27 09:00:55 ERROR 404: Not Found.
```

I also found that the GitHub workflow generated `.publish.attestation` files, but I think that they have not been uploaded:
https://github.com/py-pdf/fpdf2/actions/runs/12350082009/job/34462193992
```
Showing hash values of files to be uploaded:
/github/workspace/dist/fpdf2-2.8.2-py2.py3-none-any.whl

SHA256: 951e26290d0fc6ab4582b0d0bbacb64716fecef0f2b9223f9178c90ec4321af7
MD5: 0b91dfbebd23a01f18b6dd1465776fa1
BLAKE2-256: eb467aae9cb2584dcac217e662ab6d4670ef4e447b73d624b6210f7155322411

/github/workspace/dist/fpdf2-2.8.2.tar.gz

SHA256: 3a2c6699c39b23b786fc6ad9fc3de5432e59f6b6383bb9ab4ce1f994a5f3e762
MD5: ff01a954528f583a6609dbc1f1c0283c
BLAKE2-256: b0540e86f986e81abad9e6b348f5176048a2aa046920d46292c42a581064d93e

/github/workspace/dist/fpdf2-2.8.2-py2.py3-none-any.whl.publish.attestation

SHA256: 3f46c364e7640fda48fb698c5323613bf9a2949eacde27361157ab3875e77a88
MD5: dd2b335efc038f425de507453b36ddd2
BLAKE2-256: c1e32d67606cf5d9dc0b4adb3a8322893453cae1abb2402741b825ac33e0685c

/github/workspace/dist/fpdf2-2.8.2.tar.gz.publish.attestation

SHA256: 9bb480ab775c2717b91aa8513b4721458d44bc87cdd8447218be98c369e838a9
MD5: 2bdcbeb856d16ba8507aeac80bccc1bf
BLAKE2-256: 6111918be987f8bab95e264cfb7925514148d90be83273da2abdaff133d3828b

Uploading distributions to https://upload.pypi.org/legacy/
Uploading fpdf2-2.8.2-py2.py3-none-any.whl
Uploading fpdf2-2.8.2.tar.gz
```
```
$ wget https://github.com/actions/python-versions/releases/download/3.13.1-12154081405/python-3.13.1-linux-22.04-x64.tar.gz.publish.attestation
2024-12-27 09:00:55 ERROR 404: Not Found.
```

I also tried to use the [pypi-attestations](https://pypi.org/project/pypi-attestations/) project/tool, but it seems to also require a `.publish.attestation` file:
```
$ python -m pypi_attestations verify --identity https://github.com/py-pdf/fpdf2/.github/workflows/release.yml@refs/branch/master fpdf2-2.8.2.tar.gz
fpdf2-2.8.2.tar.gz.publish.attestation is not a file.
```

Hence my question: **why not `.publish.attestation` files have been uploaded to Pypi from my `pypa/gh-action-pypi-publish` job, and is it normal?**

**Comment by woodruffw** [comment]
> I also found that the GitHub workflow generated `.publish.attestation` files, but I think that they have not been uploaded:

They were uploaded, but they're not served at `{url}.publish.attestation`. The discovery scheme for provenance is documented here: https://docs.pypi.org/api/integrity/

TL;DR: They are uploaded to PyPI, but you need to access them via the simple HTML index (PEP 503) or JSON API (PEP 691) -- you can't just hit a URL and fetch them, since they get merged together 🙂

**Comment by Lucas-C** [comment]
Alright, thank you very much for your answer 👍

Are there plans for the `sigstore verify github` / `sigstore verify` commands to perform signature retrieval from Pypi?

I'm thinking about a command like:

    sigstore verify github fpdf2-2.8.2.tar.gz --repository=py-pdf/fpdf2 --provenance-from-pypi fpdf2

**Comment by woodruffw** [comment]
> Are there plans for the `sigstore verify github` / `sigstore verify` commands to perform signature retrieval from Pypi?

Nobody else has raised it, but I really like the idea of having this available from the `sigstore` CLI 🙂 

I'll open up a separate tracking issue on `sigstore-python` for that -- the other maintainers will likely have opinions/thoughts on it.

**Comment by Lucas-C** [comment]
Awesome @woodruffw, thank you!

I tested retrieving the singature/provenance file from Pypi, but it seems that it does not produce a valid `.sigstore.json` file:
```
$ curl https://pypi.org/integrity/fpdf2/2.8.2/fpdf2-2.8.2.tar.gz/provenance > fpdf2-2.8.2.sigstore.json

$ sigstore verify github fpdf2-2.8.2.tar.gz --repository=py-pdf/fpdf2 --sha b9cfbb6d8ca1eb034e826fd358194e899a1daf28
[21:11:19] ERROR    An issue occurred while parsing the Sigstore bundle.                                        errors.py:41

                    The provided bundle is malformed and may have been modified maliciously.

                    Additional context:

                    unsupported bundle format:

                    For detailed error information, run sigstore with the `--verbose` flag.
```

**Comment by woodruffw** [comment]
Yeah, sorry for the confusion there -- that's because the `/provenance` is a JSON blob that contains one or more attestations (just one in your case).

The CLI interfaces here are pretty shoddy at the moment, but I'm working on improving those.

The Sphinx maintainers hit a similar set of rough edges; the resolution is here: https://github.com/trailofbits/pypi-attestations/issues/55#issuecomment-2406897303

**Comment by Lucas-C** [comment]
For the record, I got _something_ working using those commands to select the 1st attestation:
```
$ curl -s https://pypi.org/integrity/fpdf2/2.8.2/fpdf2-2.8.2.tar.gz/pr
ovenance | jq -r .attestation_bundles[0].attestations[0] > fpdf2-2.8.2.tar.gz.publish.attestation

$ python -m pypi_attestations verify --identity https://github.com/py-pdf/fpdf2/.github/workflows/continuous-integration-workflow.yml@refs/heads/master fpdf2-2.8.2.tar.gz
OK: fpdf2-2.8.2.tar.gz.publish.attestation
```

It requires [`jq`](https://jqlang.github.io/jq/).

Regarding `fpdf2`, I'm going to wait for a new `sigstore verify github --provenance-from-pypi` parameter before adding signature verification in our documentation, because I'm not really what to do in case of multiple `attestation_bundles` and/or `attestations`.

**Comment by woodruffw** [comment]
Glad to hear you got it working! Sorry for the confusion there.

> I'm going to wait for a new sigstore verify github --provenance-from-pypi parameter before adding signature verification in our documentation, because I'm not really what to do in case of multiple attestation_bundles and/or attestations.

Makes sense. FWIW in the case of multiple attestations you can do one-of-N, and that's what `sigstore-python` itself will likely do: multiple attestations aren't meant to contend with each other; they're solely a way for multiple different parties to attest to different things about the package.

**Comment by Lucas-C** [comment]
> @Lucas-C we haven't cut a release for it yet, but the CLI changes here ([trailofbits/pypi-attestations#82](https://github.com/trailofbits/pypi-attestations/pull/82)) should make it much easier to experiment with provenance JSON that's been pulled from PyPI.

Thank you @woodruffw!

That is a very handy sumbcommand, thanks for introducing it 👍

I added a short paragraph to our documentation to give some pointers to our users on how to use `pypy-attestations`:
https://github.com/py-pdf/fpdf2/pull/1331

Could you have a quick look and tell me if what I wrote in this paragraph is correct?

Moreover, I have another feature request: would it be possible to introduce extra parameters to the `pypi-attestations verify pypi` subcommand in order to verify more claims, like the fact the the `fpdf2-2.8.2.tar.gz` archive has been built based on the [commit b9cfbb6](https://github.com/py-pdf/fpdf2/commit/b9cfbb6d8ca1eb034e826fd358194e899a1daf28) from our Github repo?

**Comment by woodruffw** [comment]
> Could you have a quick look and tell me if what I wrote in this paragraph is correct, please?

Those steps look correct to me!

> * instead of provided a full URL pointing to `pythonhosted.org`, could it be possible to specify `pypi-attestations verify pypi fpdf2/#fpdf2-2.8.2.tar.gz`, which would be resolved as [pypi.org/project/fpdf2#fpdf2-2.8.2.tar.gz](https://pypi.org/project/fpdf2/#fpdf2-2.8.2.tar.gz), and then to the corresponding `pythonhosted.org` URL?

> Could/should I open issues for those feature requests on [trailofbits/pypi-attestations](https://github.com/trailofbits/pypi-attestations?rgh-link-date=2025-01-13T14%3A59%3A21Z)?

Yes, please do!

A larger design pressure here is that we're trying to avoid inventing new syntax and we *don't* want `pypi-attestations` to be an official first-party CLI -- it's intended for experimentation  while we work out a larger integration plan for packaging workflows. But filing those issues will help us better understand how best to do that integration 🙂

**Comment by Lucas-C** [comment]
> A larger design pressure here is that we're trying to avoid inventing new syntax and we _don't_ want `pypi-attestations` to be an official first-party CLI -- it's intended for experimentation while we work out a larger integration plan for packaging workflows. But filing those issues will help us better understand how best to do that integration 🙂

Alright, I understand.

I opened those issues:
* https://github.com/trailofbits/pypi-attestations/issues/84
* https://github.com/trailofbits/pypi-attestations/issues/85
